packages/alfa-aria/src/role.ts

```diff
@@ -1,453 +1,456 @@
+  }
+
+  /**
+   * Get the required parent of this role.
+   */
+  public get requiredParent(): Iterable<Role.Name> {
+    return Roles[this._name].parent.required;
   }
 
-  public isSupported(predicate: Predicate<Attribute>): boolean {
-    return some(this.supported(), predicate);
+  /**
+   * Get the required children of this role.
+   */
+  public get requiredChildren(): Iterable<Iterable<Role.Name>> {
+    return Roles[this._name].children.required;
   }
 
-  public isAllowed(predicate: Predicate<Attribute>): boolean {
-    return this.isRequired(predicate) || this.isSupported(predicate);
+  /**
+   * Check if this role has the specified name.
+   */
+  public hasName<N extends Role.Name>(name: N): this is Role<N> {
+    return this._name === (name as Role.Name);
   }
 
-  public hasContext(predicate: Predicate<Role> = () => true): boolean {
-    return (
-      some(this.characteristics.context, (name) =>
-        Role.lookup(name).some(predicate)
-      ) || this.inheritsFrom((role) => role.hasContext(predicate))
-    );
+  /**
+   * Check if this role is a superclass of the role with the specified name.
+   */
+  public isSuperclassOf<N extends Role.Name>(
+    name: N
+  ): this is Role<Role.SuperclassOf<N>> {
+    const { inherited } = Roles[name];
+
+    for (const parent of inherited) {
+      if (parent === this._name || this.isSuperclassOf(parent)) {
+        return true;
+      }
+    }
+
+    return false;
   }
 
-  public hasNameFrom(predicate: Predicate<"contents" | "author">): boolean {
-    return some(this.characteristics.name.from, predicate);
+  /**
+   * Check if this role is a subclass of the role with the specified name.
+   */
+  public isSubclassOf<N extends Role.Name>(
+    name: N
+  ): this is Role<Role.SubclassOf<N>> {
+    return Role.of(name).isSuperclassOf(this._name);
   }
 
-  public hasName(predicate: Predicate<string>): boolean {
-    return predicate(this.name);
+  /**
+   * Check if this role either is, or is a subclass of, the role with the
+   * specified name.
+   */
+  public is<N extends Role.Name>(
+    name: N
+  ): this is Role<N | Role.SubclassOf<N>> {
+    return this.hasName(name) || this.isSubclassOf(name);
+  }
+
+  /**
+   * Check if this role is abstract.
+   */
+  public isAbstract(): this is Role<Role.Abstract> {
+    return Roles[this._name].abstract;
+  }
+
+  /**
+   * Check if this role is non-abstract.
+   */
+  public isConcrete(): this is Role<Role.Concrete> {
+    return !this.isAbstract();
+  }
+
+  /**
+   * Check if this role is presentational.
+   */
+  public isPresentational(): this is Role<Role.Presentational> {
+    return this.hasName("presentation") || this.hasName("none");
+  }
+
+  /**
+   * Check if this role is a widget.
+   */
+  public isWidget(): this is Role<Role.Widget> {
+    return this.is("widget");
+  }
+
+  /**
+   * Check if this role is a landmark.
+   */
+  public isLandmark(): this is Role<Role.Landmark> {
+    return this.is("landmark");
+  }
```

Comment:
I intended to include functions like these for all the different categorisations: https://www.w3.org/TR/wai-aria/#roles_categorization. The `Role#isWidget()` method actually isn't quite aligned with the "widget" categorisation and so might warrant further refinement if it turns out to cause issues.

Reply:
Make sense.
I feel there is two layers in the roles definition. One is the whole structure of roles: a thingie with sub/super classes, required context, supported attributes, … The other is a particular instance of that structure where we have a bunch of actual roles.
It is possible to define a full new roles hierarchy with actual roles being totally different from ARIA. This other hierarchy would still require `isSuperClass` or `isSupported` but wouldn't have `isLandmark` because maybe it wouldn't have a role called "landmark" in the first place.
That's why I feel maybe `isLandmark` and friends could be move to a separate layer of code: they only make sense for the particular hierarchy of roles that we consider here.

Reply:
> That's why I feel maybe isLandmark and friends could be move to a separate layer of code: they only make sense for the particular hierarchy of roles that we consider here.

That's the _only_ role hierarchy this code is intended to support though so the above sounds like a case of excess abstraction 🤔